tests/test_packages/test_skills/test_generic_buyer/test_behaviours.py

```diff
@@ -0,0 +1,349 @@
+            assert end_state_numbers == 0
+        for (
+            end_state,
+            end_state_numbers,
+        ) in fipa_dialogues.dialogue_stats.self_initiated.items():
+            if end_state == FipaDialogue.EndState.DECLINED_CFP:
+                assert end_state_numbers == 1
+            else:
+                assert end_state_numbers == 0
+
+    def test_fipa_handler_handle_decline_decline_accept(self):
+        """Test the _handle_decline method of the fipa handler where the end state is decline_accept."""
+        # setup
+        fipa_handler = cast(GenericFipaHandler, self.skill.skill_context.handlers.fipa)
+
+        incoming_message = self.build_incoming_message(
+            message_type=FipaMessage,
+            dialogue_reference=("1", "1"),
+            message_id=4,
+            target=3,
+            performative=FipaMessage.Performative.DECLINE,
+        )
+
+        fipa_dialogue = FipaDialogue(
+            DialogueLabel(("1", "1"), "counterparty", "me"),
+            "me",
+            FipaDialogue.Role.BUYER,
+        )
+        fipa_dialogues = self.skill.skill_context.fipa_dialogues
+
+        # before
+        for end_state_numbers in fipa_dialogues.dialogue_stats.self_initiated.values():
+            assert end_state_numbers == 0
+        for end_state_numbers in fipa_dialogues.dialogue_stats.other_initiated.values():
+            assert end_state_numbers == 0
+
+        # operation
+        fipa_handler._handle_decline(
+            cast(FipaMessage, incoming_message), fipa_dialogue, fipa_dialogues
+        )
+
+        # after
+        for end_state_numbers in fipa_dialogues.dialogue_stats.other_initiated.values():
+            assert end_state_numbers == 0
+        for (
+            end_state,
+            end_state_numbers,
+        ) in fipa_dialogues.dialogue_stats.self_initiated.items():
+            if end_state == FipaDialogue.EndState.DECLINED_ACCEPT:
+                assert end_state_numbers == 1
+            else:
+                assert end_state_numbers == 0
+
+    def test_fipa_handler_handle_match_accept(self):
+        """Test the _handle_match_accept method of the fipa handler."""
+        # setup
+        fipa_handler = cast(GenericFipaHandler, self.skill.skill_context.handlers.fipa)
+        strategy = cast(GenericStrategy, self.skill.skill_context.strategy)
+        strategy._is_ledger_tx = False
+
+        fipa_dialogues = cast(FipaDialogues, self.skill.skill_context.fipa_dialogues)
+        fipa_dialogue = self.build_dialogue(
+            dialogues=fipa_dialogues,
+            messages=(
+                (FipaMessage.Performative.CFP, {"query": "some_query"}),
+                (FipaMessage.Performative.PROPOSE, {"proposal": "some_proposal"}),
+                (FipaMessage.Performative.ACCEPT, {}),
+                (
+                    FipaMessage.Performative.MATCH_ACCEPT_W_INFORM,
+                    {"info": {"address": "some_term_sender_address"}},
+                ),
```

Comment:
Why four messages; why not just provide the last? Then you might have to use the private methods of the dialogue models but that's much preferable than unnecessarily bloating the test here...

Reply:
Not possible to just add the last message in a dialogue. Dialogue's structure does not allow an arbitrary population of messages (e.g. ACCEPT as a first message).